1.12 port

thommy101 opened the pull request:
Fixed almost all the issues with the 1.12 port except the new crafting system.
Changed all changed methods to the new ones.

Opened against 1.11 branch as there isn't a 1.12 yet!

Warning for the enthusiastic people, this won't compile at all!

thiakil commented on src/main/java/mekanism/client/render/tileentity/RenderChemicalCrystallizer.java:
```diff
@@ -16,7 +16,7 @@
 
 	@Override
-	public void renderTileEntityAt(TileEntityChemicalCrystallizer tileEntity, double x, double y, double z, float partialTick, int destroyStage)
+	public void func_192841_a(TileEntityChemicalCrystallizer tileEntity, double x, double y, double z, float partialTick, int destroyStage, float p_192841_10_)
```
wut?? Update MCP mappings to something post-1.12 release, like `snapshot_20170615`

thommy101 replied:
There's a reason it's work in progress ;) Updating it is on the list

thommy101 commented:
Updated the mappings, crafting is still not implemented

thommy101 commented:
Current state:

![Mekanism 1.12](https://user-images.githubusercontent.com/1036527/27253043-064ebd5e-536c-11e7-9055-aa86dfc8258e.png)

It compiles, no recipes enabled at the moment.

thommy101 commented:
Update: all non nbt related recipes are converted to JSON (no, absolutely not by hand). There are still some issues with Forge loading them, looks like none works, even of other mods. This isn't on a branch yet, will be as soon as some recipes are working.

Still on the todo list:
Fix recipes with NBT like empty/filled gas tanks
Fix IMC messages (runtime registration of recipes)
Fix JEI integration as all recipe handlers have changed

61352151511 commented:
Recipes do load just not as a compiled jar file. So in a development environment the recipes should work. NBT recipes should also work as the existing IRecipe, register it the same way you register a block/item with GameRegistry.register

thommy101 commented:
They even wont load in a development environment, tested with the one probe (one i had a dev environment for) and those are also not loading..

NBT recipes should work with IRecipe, but I want them changed to JSON as much as possible and using it's factories when needed.

61352151511 commented:
With my random utilities mod (view github) all the json recipes load in development. There's also an issue on the forge github saying they specifically don't load in the jar https://github.com/MinecraftForge/MinecraftForge/issues/4020 what does happen is after the first failed file, the rest do not attempt to load.

thommy101 commented:
Wierd.. My versions aren't working.. I'll grab a copy of yours and test if that does work over here.

61352151511 commented:
We've determined that the recipes do work, just not in the compiled environment (Known Forge bug).

thommy101 commented:
Thanks to @61352151511, first pass of recipes is working, next part are the blocks that can be disabled through config. Pull request will be updated soon.

thommy101 commented:
Another update:
All crafting recipes are ported to JSON, including NBT support.
After code cleanup of debug stuff I'll push it to this pull request.

The todo list (after cleanup):
Fix items not maintaining NBT (energy stored, owner, upgrades) on upgrade to higher tier (Already know how to fix)
Fix IMC messages
Fix JEI integration

thommy101 commented:
Ported 99% of all recipes to JSON.
They should remain NBT data, not tested as something is messed up with owner not being set (probably something NBT related I messed up)

TODO:
JEI integration
IMC messages
Testing, a lot of testing

thommy101 commented:
Some "unexpected" issue came up, Forge changed block/item registrations.

Updated to the new registry standards.
Moved oredictionary registration back to init as that's supposed to be there, if other mods have issues with it, they need to solve it.
It won't compile cleanly as I haven't upgraded the buildcraft API.

As of now I've got some more free time, so I expect a working version soon. (No, no ETA available)

TODO:
- Investigate a bug with NBT being ignored on crafting.
- JEI integration
- IMC messages, check out what's still valid to use, as crafting registration is now reloadable and maybe in the future will be send from server to client to sync
- Make some crafting elements configurable (disable crafting of some items)
- Bughunting

thommy101 commented:
NBT being ignored is a Forge bug, reported to Forge

thommy101 commented:
And just as I wanted to push some fixes, a new Forge build comes up and destroys rendering..

thommy101 commented:
Another update:
- Fixed NBT recipes, all should be working.
- Reenalbed JEI, still uses some outdated API calls, but it works. Should be updated, but won't be part of this pull request.
- Removed nonfunctional code in IMCHandler, Forge currently won't allow runtime addition/deletion of recipes.

TODO:
- NBT items still have some issues with remembering settings. Owner/input/output won't save
- Make some crafting elements configurable (disable crafting of some items)
- Bughunt!
- Update RF API which will be standalone soon, no more compiling it. There will be a maven repository of it provided by the COFH team.
- Update BuildCraft API as it won't compile in current state.

thommy101 commented:
I could use some help with the NBT stuff. If someone has an idea why stuff won't save their state, NBT whatever.

How to reproduce the bug:
Place an basic smelting factory, see that the username is unknown (no owner)
Try to open any tab, fail to open.
Try to switch security mode, fail to switch.

Looks like something isn't syncing between client and server, as the items do have a owner defined when debugging, but stuff still acts out wrong.

thommy101 commented:
Fixed the bug... Result of a renaming action in one of the first commits...

The TODO list shrinks every day! :D 
- Make some crafting elements configurable (disable crafting of some items)
Bughunt!
- Update RF API which will be standalone soon, no more compiling it. There will be a maven repository of it provided by the COFH team.
- Update BuildCraft API as it won't compile in current state.

thommy101 commented:
Updated the RF API to the standalone version and added a hard dependency on it. Could be made optional, but that requires some work that I think is outside the scope of this pull request.
Changed some bits in the BC API so it will compile.

Don't know why Travis is complaining, it compiles just fine on on my pc.

So, general call for everyone who wants to test, go ahead and please report 1.12 bugs only here untill it's merged!

aidancbrady commented:
Is this feature-complete? If so I'll go ahead and make a 1.12 branch :)

thommy101 commented:
It is feature complete, I only want to do one more commit to update it to some newer mappings and the latest recommended build of Forge. 

On the 1.12 branch, I cannot make a new branch, only the owner and ones with write permission can do that. If you make a branch of the 1.11 one I can retarget this one to that branch :) 

Tomorrow I'll push the latest commit and it'd ready for merging.

thiakil commented on build.gradle:
```diff
@@ -91,6 +95,7 @@ dependencies {
     compile "inventory-tweaks:InventoryTweaks:${config.invtweak_version}:api" // Not updated yet
     compile "li.cil.oc:OpenComputers:MC1.11.2-${config.opencomputers_version}:api" // Not updated yet
     deobfCompile "mcp.mobius.waila:Hwyla:${config.hwyla_version}"
+    deobfCompile "cofh:RedstoneFlux:${config.minecraft_version}-${config.redstoneflux_version}:universal"
```
I think you can use plain `compile` here can't you? Their API shouldn't need to be deobfed

thommy101 replied:
I'm just following the "guide" TeamCOFH has put together.
https://minecraft.curseforge.com/projects/redstone-flux

They had compile at first, but changed it to deobfcompile some time later as it caused some issues. Don't know which.

thiakil replied:
Fair enough, perhaps the example implementations.

thiakil commented on src/main/java/mekanism/common/util/MekanismUtils.java:
```diff
@@ -1493,7 +1489,7 @@ public static boolean hasUsableWrench(EntityPlayer player, BlockPos pos)
 			}
 			
-			if(isCoFHHammer(tool.getItem()) && ((IToolHammer)tool.getItem()).isUsable(tool, player, pos))
+			if(isCoFHHammer(tool.getItem())) // TODO Implement CoFH Hammer && ((IToolHammer)tool.getItem()).isUsable(tool, player, pos))
```
If IToolHammer isn't in the RedstoneFlux api, I'd say just leave that one class in `src/java/api`; as long as you don't have a copy of `package-info` and the `@Optional` is present on the Mek wrench it should all be fine

thommy101 replied:
It's now not part of the RedstoneFlux API, but from one of the other mods. Letting it stay in the api folder isn't going to work either as it's now part of another mod with probably another name. If support is needed, we need to soft depend on one of the other mods.

thiakil replied:
Nah, still the same api package id. `Optional` handles the id from package-info like a modid. All they really did was take the RF API out of the main API

https://github.com/CoFH/CoFHCore/tree/1.12/src/main/java/cofh/api/item

Shouldnt really need a dep for it.

thommy101 replied:
Aaah, that package wasn't available when I had written it. I did write soft dependency, I ment optional dependency ;)

For this I'll let the COFH hammer out of this, if it's really needed it will come back in the future.

thiakil replied:
Yeah their GH is a bit weird. Like the api stuff was in CofhLib, which was a submodule of Core, and then got packaged into core, and never published as an api jar like it should have

thiakil commented:
Travis seems to have some weird warnings with the guava lib, perhaps it's cache is dirty or it has a different version that your local. Or could even be the java version I guess.

As for creating a 1.12 branch, it could easily be done manually by Aidan.

```
git remote add thommy101 https://github.com/thommy101/Mekanism.git
git checkout -b 1.12 remotes/thommy101/1.12_firstpass
git push -u origin 1.12:1.12
```

thiakil commented:
> update it to some newer mappings and the latest recommended build of Forge.

Something to note - depending on how new the mappings are, some `AxisAlignedBB` methods swapped names; the code will still compile if used, but they won't do the same thing!

thommy101 commented:
There are some issues with Guava 21 related to the annotations not being available, which is a known issue and fixed in version 22. It does compile, only Travis is having some "issues" with it.

It will probably be fixed by moving to ConnectedTexturesMod, a mod split of from Chissel which is being used for rendering the CTMs. But that's something for another pull request sometime after this merge ;)

I'll take a look at the AABB @thiakil mentioned to make sure they still have the correct names.

thommy101 commented:
Updated to 1.12 recommended, fixed some missed SideOnly changes (getSubBlocks is now callable on serverside).

Checked the AABB @thiakill mentioned, all work still as intended.

@aidancbrady It's ready for a merge on a 1.12 branch

thommy101 commented:
\* I should test compile before committing *

61352151511 commented:
Compiled from github and put it on a server, when launching I get a crash, launches fine on client. I am aware there are other mods installed however I feel it will probably happen with just mekanism as well (Feel free to test it.) https://pastebin.com/Cbgm7ixu

thiakil commented:
@thommy101 the `NoClassDefFoundError` error listed in that crash log is rather misleading; it's actually failing to load some Mekanism `Item` classes.

Caused by a few places missing `@SideOnly(Side.CLIENT)` on `public void addInformation(ItemStack itemstack, World world, List<String> list, ITooltipFlag flag)` (ITooltipFlag is client only)

If you're still getting this after that, set a breakpoint in `net.minecraftforge.fml.common.asm.transformers.SideTransformer L62` and trace back the stack until `name` contains something Mekanism. (NB you will get an exception in FMLSanityCheck, this is normal)

thommy101 commented:
Thanks for the debugging @thiakil!
Will fix today ;)

thommy101 commented:
@61352151511 Fixed!

@aidancbrady I'll add a form of crafting disabling to this pull request, It won't sync as the current Forge system initialises the crafting recipes on start-up, not on server join after a config sync as implemented by Mekanism. First pass of it is done (not yet committed) but will follow later today.

thommy101 commented:
Crafting recipes can be disabled once again, config has remain unchanged, init of the config has moved to preinit as recipe loading happens between preinit and init. This move has no impact on anything.

I've edited https://github.com/aidancbrady/Mekanism/pull/4472#issuecomment-313629832 with the current state.

aidancbrady commented:
Really appreciate the work being put into this.

@thommy101 and @thiakil, would you be interested in joining a slack team or GroupMe to discuss Mekanism development? With all the other work I have on my plate right now it'd be a much easier way to get in contact with me. Besides, as great is GitHub is for collaborative development, it's not the best messaging platform.

thiakil commented:
I'd be happy to help if you'd find that useful :)

aidancbrady commented:
Great. I've just created a new slack team -- mekanismmod.slack.com. If you can give me your email (either here or send me an email at me@aidancbrady.com) I'll add you :)

thommy101 commented:
Done and thanks! :)

61352151511 commented:
Can not craft solar generator, unsure as to why.

thommy101 commented:
Found the source of the issue, fixed in a couple minutes! Thanks for testing and reporting :D

thiakil commented on src/main/java/mekanism/common/integration/IMCHandler.java:
```diff
@@ -24,105 +21,55 @@ public void onIMCEvent(List<IMCMessage> messages)
-					
-					String message = msg.key;
-					
-					if(message.equals("ShapedMekanismRecipe"))
```
I was thinking about this tonight, perhaps we should log an error for the unlikely case that someone's using this? Rather than just silently failing

thommy101 replied:
Fixed, logging a warning currently when one of the 4 messages is being received.

BionicFelps commented:
Not sure if this is in older versions of Mekanism as well, I suspect it is, but I was trying with this.
Placing a forestry water can or capsule in an electrolytic separator or evaporator (or probably anything that'll accept water from a bucket) using the gui causes a crash.

Log: https://paste.ee/p/GNykf

thiakil commented:
Crafting formula seems to be missing a model for me

thommy101 commented:
Weird, it's working in 1.11 and not in 1.12 and I haven't touched any of the code which put it in the creative tab or alter it's textures...

thiakil commented:
It's likely something to do with its model registration I guess?

thommy101 commented:
Looks like it, I guess there's some code deleted which shouldn't be deleted

thommy101 commented:
Found it! It's models are being registered in pre init instead of the newly added render event. Currently in pre-init the blocks aren't registered yet. That happens between pre-init and init.

thommy101 commented:
Most where already in the right place luckily, fixed it :)

thommy101 commented:
And updated with upstream to have the latest changes in here. The Travis warning will be fixed after #4548 is done and merged :)

aidancbrady commented:
Closing- everything has moved to the 1.12 branch.
